PM Fixes and leftovers

ceolin opened the pull request:
- ~~Send state changes notifications after SoC post operations~~
  - Make pm_system_suspend() private. This function is not intended to be called directly.
 - Effectively allow forcing a state

Commit:
power: Make pm_system_suspend private

pm_system_suspend is called only from the idle thread and should
not be exported as a public API.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Effectively allow forcing a state

pm_power_state_force had two different behavior, if
CONFIG_PM_DIRECT_FORCE_MODE was enabled this function immediately
calls pm_system_suspend() without wait the idle thread. Without this
option enabled, this function will wait for the idle thread run but
will use the given power state instead of asking the policy manager.

The problem in both cases is that the process involves handling
devices and the way that was implemented if at least one device failed
to go to low power or suspended the system power state would not
change / be forced.

This commit simplifies this API removing the conditional behavior
since it is not clear the need for that and effectively, and
immediately, forces the system to go to the given state without
bother with devices.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: s/POWER_STATE_ACTIVE/PM_STATE_ACTIVE

Fix some references to old power state names.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Remove unnecessary pm_system_mgr function

There is no need for this function. This function is called only in
one place and the code can be moved to the function that was calling
it.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

nashif commented on kernel/include/kernel_internal.h:
```diff
@@ -207,6 +207,27 @@ void z_kernel_map_fixup(void);
 			!__i.key;					\
 			k_spin_unlock(lck, __key), __i.key = 1)
 
+#ifdef CONFIG_PM
```
first, your commit message is missing a NOT, second, can we put this in some pm specific private header instead?

nashif commented on subsys/power/power.c:
```diff
@@ -208,12 +208,6 @@ static enum pm_state pm_policy_mgr(int32_t ticks)
 }
 
-
-enum pm_state pm_system_suspend(int32_t ticks)
```
this was added by me as a placeholder, idle thread is supposed to call the the policy manager  and remove `pm_system_suspend` which is misleading. `pm_system_suspend` being called in `idle()` gives the impression that we are going to suspend everytime when in fact, the policy manager decided which state we transition to.

So I would remove the `pm_system_suspend()` from `idle()` and call policy manager directly and adapt in other places as well.

ceolin replied:
that is fine, and makes the other commit not necessary. Let me re-work this.

ceolin replied:
actually thinking a little bit more about it we have two possibilities here:
1) idle thread -> policy manager  -> power subsystem
2) idle thread -> power subsystem -> policy manager

We are currently doing 2 but taking the approach 1 may simplify some things, let me change it to check

ceolin replied:
Hey @nashif I have looked it more carefully and there are some drawbacks making the idle thread talk directly with the policy. First, pm_system_suspend will need be public, because all policies will need to call it, and our model allows application create their own policies. This api is not intend to be public at all. The other reason is that I think we want to keep the policy simple, just returning the next state that the system should go and the power subsystem is responsible for the rest. We can discuss it, but I prefer discuss it before do the changes.

Regarding the header, the problem with creating a `power_internal.h` is that this header is not going to be use in the power subsystem but in the kernel. I didn't see other subsystems doing it. Don't know how is the best way to organize it. The situation is that we have a functionality that is implemented by this subsystem that is only used by the kernel, not even the subsystem itself needs to call it, that is similar of `arch_interface.h` but for this subsystem.

Anyway, I sent a new version fixing the commit message and adding more cleanup, please review it again and let me know about the header. The `pm_system_suspend` call from the idle thread is something we can discuss more, this commit is not introducing it there, just simplifying.

Commit:
power: Remove unnecessary pm_idle_exit_notification_disable api

This function is useless and the state variable that it was
controlling is also not necessary because the same logic is being
handled by the variable post_ops_done.\

This reasonably simplifies idle thread logic.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Remove unused / unimplemented code

pm_system_resume_from_deep_sleep is not implemented or used
anywhere. Just remove it and keep the code base cleaner.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
idle: Remove weak function

pm_system_resume is always implemented when PM is enabled. There is no
need to have this weak function under an ifdef PM.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Make pm_system_resume private

This API is not intended to be public and it is called only from the
idle thread.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Send state changes notification after device PM

Currently the power subsys is sending notifications about power state
changes before doing device pm. If one device fails to low power or
suspend the system never goes to an idle state. Change to send
notifications just before call SoC to suspend and avoid misleading
information.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Get rid of deep sleep and sleep concepts

New power states have more granularity than deep sleep and sleep
states. Just get rid of this and keep the same behavior for now.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
@nashif  ping :)

I've removed one commit that was changing the order that state changes notification were being sent to try to get it merged. That commit introduced a race condition that I've to think a little bit more about.

nashif commented on subsys/power/power.c:
```diff
@@ -144,7 +144,7 @@ static enum pm_state _handle_device_abort(struct pm_state_info info)
 	return PM_STATE_ACTIVE;
 }
 
-static enum pm_state pm_policy_mgr(int32_t ticks)
```
the commit is talking about `pm_system_mgr`
